lib/on_form/saving.rb

```diff
@@ -14,26 +14,21 @@ def transaction(&block)
       end
     end
 
-    def invalid?
-      !valid?
-    end
 
     def save!
       reset_errors
       transaction do
         reset_errors
-        unless run_validations!(backing_model_validations: false)
-          backing_model_instances.each(&:valid?)
-          collect_errors
-          raise ActiveModel::ValidationError, self
-        end
+
+        self.backing_model_validations = false
+        has_form_errors = !run_validations!
+        self.backing_model_validations = true
+        run_backing_model_validation!
+        raise ActiveModel::ValidationError, self if has_form_errors
+        raise ActiveRecord::RecordInvalid, self unless errors.empty?
+
         run_callbacks :save do
-          begin
-            backing_model_instances.each { |backing_model| backing_model.save! }
-          rescue ActiveRecord::RecordInvalid, ActiveModel::ValidationError
-            collect_errors
-            raise
-          end
+          backing_model_instances.each { |backing_model| backing_model.save(validate: false) }
```

Comment:
What happens now if a before_save callback on the backing model saves another model with a save! call?

Reply:
Interesting point. I guess my intuitive idea of how it would work is

* validations (+ callbacks) on the form object
  * validations (+ callbacks) on directly associated models
    * validations (+ callbacks) on nested models

then assuming these pass ...

* in a transaction
  * save callbacks on the form objects
    * save callbacks on directly associated models
      * save callbacks on nested models

So if a before_save callback on the backing model calls `save!` and it explodes, the transaction would roll back and nothing would be saved.

Not sure how well that maps to what ActiveRecord does.